.mage/end_to_end.go

```diff
@@ -0,0 +1,163 @@
+// Copyright © 2020 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ttnmage
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/magefile/mage/mg"
+	"github.com/magefile/mage/target"
+)
+
+// EndToEnd namespace
+type EndToEnd mg.Namespace
+
+var (
+	testDatabaseName = "ttn_lorawan_test"
+	databaseURI      = "postgresql://root@localhost:26257/" + testDatabaseName + "?sslmode=disable"
+)
+
+func getTestURL() string {
+	nodeEnv := os.Getenv("NODE_ENV")
+	port := "1885"
+	if nodeEnv == "development" {
+		port = "8080"
+	}
+	return "http://localhost:" + port
+}
+
+func (endToEnd EndToEnd) cypress() (func(args ...string) error, error) {
+	return execFromNodeBin("cypress", true, "--config-file", "./config/cypress.json", "--config", "baseUrl="+getTestURL())
+}
+
+func (endToEnd EndToEnd) prepareDB() error {
+	isCI := os.Getenv("CI") == "true"
+	dumpExists := pathExists("./.cache/sqldump.sql")
+	changed, err := target.Dir("./.cache/sqldump.sql", "./pkg/identityserver/store")
+	if err != nil {
+		return err
+	}
+	if (changed && !isCI) || !dumpExists {
+		os.Setenv("TTN_LW_IS_DATABASE_URI", databaseURI)
+		mg.SerialDeps(Dev.DBErase, Dev.InitStack, EndToEnd.DBDump)
+	} else {
+		mg.SerialDeps(EndToEnd.DBRestore)
+	}
+	return nil
+}
+
+func (endToEnd EndToEnd) prepareBuild() {
+	mg.SerialDeps(Js.Deps, Js.BuildIfNecessary)
```

Comment:
This is incorrect - `Js.BuildIfNecessary` should have a dependency on `js.Deps` if it needs `Js.Deps`, otherwise this should not be `SerialDeps`